[Inductor] Fix constant folding and extern kernel mutation tracking bugs

oulgen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115291
* __->__ #115908

This PR fixes two bugs
1) Constant folding a triton kernel results in the kernel's inputs to be returned back without any modification. Disable constant folding for triton kernels. Need more investigation
2) NoneLayout buffers should not be deleted as they do not exist



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

oulgen committed:
[Inductor] Fix constant folding and extern kernel mutation tracking bugs

[ghstack-poisoned]

aakhundov reviewed (approved):
Thanks for the fix!

oulgen committed:
Update on "[Inductor] Fix constant folding and extern kernel mutation tracking bugs"


This PR fixes two bugs
1) Constant folding a triton kernel results in the kernel's inputs to be returned back without any modification. Disable constant folding for triton kernels. Need more investigation
2) ExternKernelOut (matmul in this case) needs to track mutation of its inputs



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

oulgen committed:
Update on "[Inductor] Fix constant folding and extern kernel mutation tracking bugs"


This PR fixes two bugs
1) Constant folding a triton kernel results in the kernel's inputs to be returned back without any modification. Disable constant folding for triton kernels. Need more investigation
2) ExternKernelOut (matmul in this case) needs to track mutation of its inputs



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison commented on torch/_inductor/constant_folding.py:
```diff
@@ -109,6 +109,13 @@ def set_env(arg):
         ):
             return self.unknown_value
 
+        # TODO - constant folding triton kernel returns the inputs -- fix this
```
`# TODO - constant folding triton kernel returns the inputs -- fix this`

What is the TODO here exactly ?

oulgen replied:
I guess it is to evaluate whether we can in fact constant fold or not. Since it is returning the inputs, i suspect the fake tensor dispatcher is being used here but intuitively, it should just be `ProxyTorchDispatchMode`

eellison replied:
Constant Folding is running the ops directly, it shouldnt be using proxy or fake tensor. 

Could we add test ? Also, do we generate a schema for these ops

oulgen replied:
The test I added to this PR was running into both constant folding and externel kernel out mutation tracking bugs.
The constant folding was happing in joint graph passes. Did you want a different test?

eellison replied:
Okay nevermind, it doesn't have a schema then.

oulgen committed:
Update on "[Inductor] Fix constant folding and extern kernel mutation tracking bugs"


This PR fixes two bugs
1) Constant folding a triton kernel results in the kernel's inputs to be returned back without any modification. Disable constant folding for triton kernels. Need more investigation
2) ExternKernelOut (matmul in this case) needs to track mutation of its inputs



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

oulgen commented:
So this PR results in matmul input buffers to live longer
![410426936_2692549810921842_1229635650480369372_n](https://github.com/pytorch/pytorch/assets/4007610/3bdc53a8-28dc-47b8-81c8-c01b3b1a2343)

the difference is
OLD
```
x = EMPTY()
y = matrix multiply (x...)
z = reinterp(x)
del x
```
NEW
```
x = prev_z
del prev_x
y = matrix_multiply(some_old_value...)
z = empty()
```
so in the new one z unnecessarily lives until the next matrix multiply

oulgen commented:
@aakhundov also discovered that
so triton_poi_fused__to_copy_add_mul_pow_tanh_backward_10  used to be:

```
def triton_(in_out_ptr0, in_ptr0, in_ptr1, xnumel, XBLOCK : tl.constexpr):
    …
    tmp0 = tl.load(in_out_ptr0 + (x0), None).to(tl.float32)
    …
    tl.store(in_out_ptr0 + (x0), tmp28, None)
```
and now it is
```
def triton_(in_ptr0, in_ptr1, in_ptr2, out_ptr0, xnumel, XBLOCK : tl.constexpr):
    …
    tmp0 = tl.load(in_ptr0 + (x0), None).to(tl.float32)
    …
    tl.store(out_ptr0 + (x0), tmp28, None)  
```

oulgen committed:
Update on "[Inductor] Fix constant folding and extern kernel mutation tracking bugs"


This PR fixes two bugs
1) Constant folding a triton kernel results in the kernel's inputs to be returned back without any modification. Disable constant folding for triton kernels. Need more investigation
2) ExternKernelOut (matmul in this case) needs to track mutation of its inputs



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

oulgen commented:
Mutation tracking looks good

![kernel](https://github.com/pytorch/pytorch/assets/4007610/90f55e9e-69f1-4dfa-8719-c456ffd295b2)

oulgen committed:
Update on "[Inductor] Fix constant folding and extern kernel mutation tracking bugs"


This PR fixes two bugs
1) Constant folding a triton kernel results in the kernel's inputs to be returned back without any modification. Disable constant folding for triton kernels. Need more investigation
2) ExternKernelOut (matmul in this case) needs to track mutation of its inputs



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

oulgen committed:
Update on "[Inductor] Fix constant folding and extern kernel mutation tracking bugs"


This PR fixes two bugs
1) Constant folding a triton kernel results in the kernel's inputs to be returned back without any modification. Disable constant folding for triton kernels. Need more investigation
2) ExternKernelOut (matmul in this case) needs to track mutation of its inputs



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

oulgen commented:
@pytorchbot merge
